pkg/common/consumer/managed_group.go

```diff
@@ -19,34 +19,107 @@ package consumer
 import (
 	"context"
 	"sync/atomic"
+	"time"
 
 	"github.com/Shopify/sarama"
+	"go.uber.org/zap"
 )
 
 // managedGroup contains information about a Sarama ConsumerGroup that is required to start (i.e. re-create) it
 type managedGroup struct {
-	group              sarama.ConsumerGroup
-	errors             chan error
-	restartWaitChannel chan struct{}
-	stopped            atomic.Value // Boolean value indicating channel is stopped
-	cancelErrors       func()
-	cancelConsume      func()
+	logger             *zap.Logger
+	group              sarama.ConsumerGroup // The Sarama ConsumerGroup which is under management
+	errors             chan error           // An error channel that will replicate the errors from the Sarama ConsumerGroup
+	restartWaitChannel chan struct{}        // A channel that will be closed when a stopped group is restarted
+	stopped            atomic.Value         // Boolean value indicating that the managed group is stopped
+	cancelErrors       func()               // Called by the manager's CloseConsumerGroup to terminate the error forwarding
+	cancelConsume      func()               // Called by the manager during CloseConsumerGroup
+	lockedBy           atomic.Value         // The LockToken of the ConsumerGroupAsyncCommand that requested the lock
+	cancelLockTimeout  func()               // Called internally to stop the lock timeout when a lock is removed
 }
 
 // createManagedGroup associates a Sarama ConsumerGroup and cancel function (usually from the factory)
-// inside a new managedGroup struct
-func createManagedGroup(group sarama.ConsumerGroup, cancel func()) *managedGroup {
+// inside a new managedGroup struct.  If a timeout is given (nonzero), the lockId will be reset to an
+// empty string (i.e. "unlocked") after that time has passed.
+func createManagedGroup(logger *zap.Logger, group sarama.ConsumerGroup, cancel func()) *managedGroup {
+
 	managedGrp := managedGroup{
+		logger:       logger,
 		group:        group,
 		errors:       make(chan error),
 		cancelErrors: cancel,
 	}
-	// Initialize the atomic boolean value to false, otherwise it will be a nil interface
-	// (restartWaitChannel is only accessed when "stopped" is true)
-	managedGrp.stopped.Store(false)
+
+	// Atomic values must be initialized with their desired type before being accessed, or a nil
+	// interface will be returned that is not of that type (string or bool in these cases)
+	managedGrp.lockedBy.Store("")   // Empty token string indicates "unlocked"
+	managedGrp.stopped.Store(false) // A managed group defaults to "started" when created
+
 	return &managedGrp
 }
 
+// resetLockTimer will set the "lockedBy" field to the given lockToken (need not be the same as the
+// existing one) and reset the timer to the provided timeout.  After the timer expires, the lockToken
+// will be set to an empty string (representing "unlocked")
+func (m *managedGroup) resetLockTimer(lockToken string, timeout time.Duration) {
```

Comment:
I would rename to `resetLock` since it doesn't only change the timer but replaces the token ?

Reply:
That's probably more succinct.  Originally I was thinking "reset lock and timer"

Reply:
that's fine too - whatever you prefer to make it clear.